File: app/src/main/java/com/example/purepawapp/ui/admin/AdminDashboardFragment.java
package com.example.purepawapp.ui.admin;

import android.graphics.Color;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.navigation.fragment.NavHostFragment;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.Order;
import com.example.purepawapp.data.model.Product;
import com.example.purepawapp.data.model.User;
import com.example.purepawapp.data.repository.RepoCallback;
import com.example.purepawapp.databinding.FragmentAdminDashboardBinding;
import com.example.purepawapp.databinding.ItemAdminOrderBinding;
import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.ui.common.BaseFragment;
import com.example.purepawapp.util.CurrencyUtils;

import java.util.ArrayList;
import java.util.Calendar;
import java.util.Comparator;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicReference;

public class AdminDashboardFragment extends BaseFragment<FragmentAdminDashboardBinding> {

    public AdminDashboardFragment() {
        super(FragmentAdminDashboardBinding::inflate);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        getBinding().tvSeeAllOrders.setOnClickListener(v -> NavHostFragment.findNavController(this).navigate(R.id.adminOrdersFragment));

        String uid = ServiceLocator.getAuthRepository().getCurrentUserId();
        if (uid != null) {
            ServiceLocator.getProfileRepository().getUser(uid, new RepoCallback<>() {
                @Override
                public void onSuccess(User user) {
                    String name = user.getFullName();
                    getBinding().tvAdminName.setText(name == null || name.isBlank() ? "Admin PurePaw" : name);
                }

                @Override
                public void onError(Exception error) {
                }
            });
        }

        loadDashboard();
    }

    @Override
    public void onResume() {
        super.onResume();
        loadDashboard();
    }

    private void loadDashboard() {
        AtomicReference<List<Product>> productsRef = new AtomicReference<>(List.of());
        AtomicReference<List<Order>> ordersRef = new AtomicReference<>(List.of());
        AtomicReference<List<User>> usersRef = new AtomicReference<>(List.of());
        AtomicInteger pending = new AtomicInteger(3);

        Runnable maybeFinish = () -> {
            if (pending.decrementAndGet() == 0) {
                List<Product> products = productsRef.get();
                List<Order> orders = ordersRef.get();
                List<User> users = usersRef.get();

                getBinding().tvProductCount.setText(String.valueOf(products.size()));
                getBinding().tvOrderCount.setText(String.valueOf(orders.size()));
                getBinding().tvUserCount.setText(String.valueOf(users.size()));
                double revenue = 0;
                for (Order o : orders) revenue += o.getTotalAmount();
                getBinding().tvRevenue.setText(AdminStatusUi.toCompactVnd(revenue));

                renderChart(orders);
                Map<String, String> userNames = new HashMap<>();
                for (User u : users) userNames.put(u.getUid(), u.getFullName());
                renderRecentOrders(orders, userNames);
            }
        };

        ServiceLocator.getProductRepository().getProducts(null, new RepoCallback<>() {
            @Override
            public void onSuccess(List<Product> result) {
                productsRef.set(result);
                maybeFinish.run();
            }

            @Override
            public void onError(Exception error) {
                maybeFinish.run();
            }
        });

        ServiceLocator.getOrderRepository().getAllOrders(new RepoCallback<>() {
            @Override
            public void onSuccess(List<Order> result) {
                ordersRef.set(result);
                maybeFinish.run();
            }

            @Override
            public void onError(Exception error) {
                maybeFinish.run();
            }
        });

        ServiceLocator.getProfileRepository().getAllUsers(new RepoCallback<>() {
            @Override
            public void onSuccess(List<User> result) {
                usersRef.set(result);
                maybeFinish.run();
            }

            @Override
            public void onError(Exception error) {
                maybeFinish.run();
            }
        });
    }

    private void renderChart(List<Order> orders) {
        getBinding().llChart.removeAllViews();
        Calendar now = Calendar.getInstance();
        List<Calendar> monthKeys = new ArrayList<>();
        for (int offset = 5; offset >= 0; offset--) {
            Calendar cal = (Calendar) now.clone();
            cal.add(Calendar.MONTH, -offset);
            monthKeys.add(cal);
        }

        List<Double> revenueByMonth = new ArrayList<>();
        for (Calendar cal : monthKeys) {
            double sum = 0;
            for (Order order : orders) {
                Calendar orderCal = Calendar.getInstance();
                orderCal.setTimeInMillis(order.getCreatedAt());
                if (orderCal.get(Calendar.YEAR) == cal.get(Calendar.YEAR) && orderCal.get(Calendar.MONTH) == cal.get(Calendar.MONTH)) {
                    sum += order.getTotalAmount();
                }
            }
            revenueByMonth.add(sum);
        }

        double maxRevenue = 1.0;
        for (double v : revenueByMonth) if (v > maxRevenue) maxRevenue = v;

        int maxBarHeightDp = 90;
        float density = getResources().getDisplayMetrics().density;

        for (int index = 0; index < monthKeys.size(); index++) {
            Calendar cal = monthKeys.get(index);
            LinearLayout column = new LinearLayout(requireContext());
            column.setOrientation(LinearLayout.VERTICAL);
            column.setGravity(Gravity.BOTTOM | Gravity.CENTER_HORIZONTAL);
            column.setLayoutParams(new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, 1f));

            int barHeightPx = Math.max((int) (maxBarHeightDp * density * (revenueByMonth.get(index) / maxRevenue)), (int) (4 * density));
            View bar = new View(requireContext());
            bar.setBackgroundColor(getResources().getColor(R.color.pp_primary, null));
            bar.setLayoutParams(new LinearLayout.LayoutParams((int) (20 * density), barHeightPx));

            TextView label = new TextView(requireContext());
            label.setText("T" + (cal.get(Calendar.MONTH) + 1));
            label.setTextSize(11f);
            label.setTextColor(getResources().getColor(R.color.pp_text_secondary, null));
            label.setGravity(Gravity.CENTER);
            label.setPadding(0, (int) (4 * density), 0, 0);

            column.addView(bar);
            column.addView(label);
            getBinding().llChart.addView(column);
        }

        if (orders.isEmpty()) {
            getBinding().llChart.removeAllViews();
            TextView empty = new TextView(requireContext());
            empty.setText("Chưa có dữ liệu doanh thu");
            empty.setTextColor(Color.parseColor("#8A7B6A"));
            empty.setTextSize(13f);
            empty.setGravity(Gravity.CENTER);
            empty.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.MATCH_PARENT));
            getBinding().llChart.addView(empty);
        }
    }

    private void renderRecentOrders(List<Order> orders, Map<String, String> userNames) {
        getBinding().llRecentOrders.removeAllViews();
        List<Order> recent = new ArrayList<>(orders);
        recent.sort(Comparator.comparingLong(Order::getCreatedAt).reversed());
        if (recent.size() > 5) recent = recent.subList(0, 5);
        getBinding().tvNoOrders.setVisibility(recent.isEmpty() ? View.VISIBLE : View.GONE);

        for (Order order : recent) {
            ItemAdminOrderBinding rowBinding = ItemAdminOrderBinding.inflate(getLayoutInflater(), getBinding().llRecentOrders, false);
            String customerName = userNames.get(order.getUserId());
            if (customerName == null || customerName.isBlank()) {
                customerName = order.getShippingAddress().getFullName();
                if (customerName == null || customerName.isBlank()) customerName = "Khách hàng";
            }
            rowBinding.tvCustomerName.setText(customerName);
            String orderCode = order.getOrderCode();
            rowBinding.tvOrderCode.setText(orderCode != null && !orderCode.isBlank() ? orderCode : "#" + order.getId().substring(0, Math.min(8, order.getId().length())));
            int itemCount = 0;
            for (var item : order.getItems()) itemCount += item.getQuantity();
            rowBinding.tvItemCount.setText(itemCount + " sản phẩm");
            rowBinding.tvTotal.setText(CurrencyUtils.toVndString(order.getTotalAmount()));
            StatusStyle style = AdminStatusUi.orderStatusStyle(order.getStatus());
            rowBinding.tvStatus.setText(style.getLabel());
            rowBinding.tvStatus.setTextColor(getResources().getColor(style.getTextColorRes(), null));
            rowBinding.tvStatus.setBackgroundTintList(getResources().getColorStateList(style.getBgColorRes(), null));
            rowBinding.getRoot().setOnClickListener(v -> NavHostFragment.findNavController(this).navigate(R.id.adminOrdersFragment));
            if (rowBinding.getRoot().getLayoutParams() instanceof LinearLayout.LayoutParams) {
                LinearLayout.LayoutParams params = (LinearLayout.LayoutParams) rowBinding.getRoot().getLayoutParams();
                params.bottomMargin = (int) (8 * getResources().getDisplayMetrics().density);
            }
            getBinding().llRecentOrders.addView(rowBinding.getRoot());
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/admin/AdminDashboardFragment.kt
package com.example.purepawapp.ui.admin

import android.graphics.Color
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import com.example.purepawapp.R
import com.example.purepawapp.data.model.Order
import com.example.purepawapp.databinding.FragmentAdminDashboardBinding
import com.example.purepawapp.databinding.ItemAdminOrderBinding
import com.example.purepawapp.di.ServiceLocator
import com.example.purepawapp.ui.common.BaseFragment
import com.example.purepawapp.util.toVndString
import kotlinx.coroutines.launch
import java.util.Calendar

class AdminDashboardFragment : BaseFragment<FragmentAdminDashboardBinding>(FragmentAdminDashboardBinding::inflate) {

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.tvSeeAllOrders.setOnClickListener {
            findNavController().navigate(R.id.adminOrdersFragment)
        }

        val uid = ServiceLocator.authRepository.currentUserId
        if (uid != null) {
            viewLifecycleOwner.lifecycleScope.launch {
                ServiceLocator.profileRepository.getUser(uid).onSuccess { user ->
                    binding.tvAdminName.text = user.fullName.ifBlank { "Admin PurePaw" }
                }
            }
        }

        loadDashboard()
    }

    override fun onResume() {
        super.onResume()
        loadDashboard()
    }

    private fun loadDashboard() {
        viewLifecycleOwner.lifecycleScope.launch {
            val products = ServiceLocator.productRepository.getProducts().getOrDefault(emptyList())
            val orders = ServiceLocator.orderRepository.getAllOrders().getOrDefault(emptyList())
            val users = ServiceLocator.profileRepository.getAllUsers().getOrDefault(emptyList())

            binding.tvProductCount.text = products.size.toString()
            binding.tvOrderCount.text = orders.size.toString()
            binding.tvUserCount.text = users.size.toString()
            binding.tvRevenue.text = orders.sumOf { it.totalAmount }.toCompactVnd()

            renderChart(orders)
            renderRecentOrders(orders, users.associate { it.uid to it.fullName })
        }
    }

    private fun renderChart(orders: List<Order>) {
        binding.llChart.removeAllViews()
        val now = Calendar.getInstance()
        val monthKeys = (5 downTo 0).map { offset ->
            (now.clone() as Calendar).apply { add(Calendar.MONTH, -offset) }
        }
        val revenueByMonth = monthKeys.map { cal ->
            orders.filter {
                val orderCal = Calendar.getInstance().apply { timeInMillis = it.createdAt }
                orderCal.get(Calendar.YEAR) == cal.get(Calendar.YEAR) &&
                    orderCal.get(Calendar.MONTH) == cal.get(Calendar.MONTH)
            }.sumOf { it.totalAmount }
        }
        val maxRevenue = revenueByMonth.maxOrNull()?.takeIf { it > 0 } ?: 1.0
        val maxBarHeightDp = 90
        val density = resources.displayMetrics.density

        monthKeys.forEachIndexed { index, cal ->
            val column = LinearLayout(requireContext()).apply {
                orientation = LinearLayout.VERTICAL
                gravity = Gravity.BOTTOM or Gravity.CENTER_HORIZONTAL
                layoutParams = LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.MATCH_PARENT, 1f)
            }
            val barHeightPx = (maxBarHeightDp * density * (revenueByMonth[index] / maxRevenue)).toInt()
                .coerceAtLeast((4 * density).toInt())
            val bar = View(requireContext()).apply {
                setBackgroundColor(resources.getColor(R.color.pp_primary, null))
                layoutParams = LinearLayout.LayoutParams((20 * density).toInt(), barHeightPx)
            }
            val label = TextView(requireContext()).apply {
                text = "T${cal.get(Calendar.MONTH) + 1}"
                textSize = 11f
                setTextColor(resources.getColor(R.color.pp_text_secondary, null))
                gravity = Gravity.CENTER
                setPadding(0, (4 * density).toInt(), 0, 0)
            }
            column.addView(bar)
            column.addView(label)
            binding.llChart.addView(column)
        }
        if (orders.isEmpty()) {
            binding.llChart.removeAllViews()
            binding.llChart.addView(
                TextView(requireContext()).apply {
                    text = "Chưa có dữ liệu doanh thu"
                    setTextColor(Color.parseColor("#8A7B6A"))
                    textSize = 13f
                    gravity = Gravity.CENTER
                    layoutParams = LinearLayout.LayoutParams(
                        LinearLayout.LayoutParams.MATCH_PARENT,
                        LinearLayout.LayoutParams.MATCH_PARENT
                    )
                }
            )
        }
    }

    private fun renderRecentOrders(orders: List<Order>, userNames: Map<String, String>) {
        binding.llRecentOrders.removeAllViews()
        val recent = orders.sortedByDescending { it.createdAt }.take(5)
        binding.tvNoOrders.visibility = if (recent.isEmpty()) View.VISIBLE else View.GONE

        recent.forEach { order ->
            val rowBinding = ItemAdminOrderBinding.inflate(layoutInflater, binding.llRecentOrders, false)
            rowBinding.tvCustomerName.text = userNames[order.userId] ?: order.shippingAddress.fullName.ifBlank { "Khách hàng" }
            rowBinding.tvOrderCode.text = order.orderCode.ifBlank { "#${order.id.take(8)}" }
            rowBinding.tvItemCount.text = "${order.items.sumOf { it.quantity }} sản phẩm"
            rowBinding.tvTotal.text = order.totalAmount.toVndString()
            val style = orderStatusStyle(order.status)
            rowBinding.tvStatus.text = style.label
            rowBinding.tvStatus.setTextColor(resources.getColor(style.textColorRes, null))
            rowBinding.tvStatus.backgroundTintList = resources.getColorStateList(style.bgColorRes, null)
            rowBinding.root.setOnClickListener {
                findNavController().navigate(R.id.adminOrdersFragment)
            }
            (rowBinding.root.layoutParams as? LinearLayout.LayoutParams)?.bottomMargin =
                (8 * resources.displayMetrics.density).toInt()
            binding.llRecentOrders.addView(rowBinding.root)
        }
    }
}
